src/Common/src/System/Numerics/Hashing/Hash.cs

```diff
@@ -0,0 +1,96 @@
+        }
+        
+        public static int Combine(int h1, int h2, int h3, int h4, int h5)
+        {
+            return Combine(Combine(h1, h2, h3, h4), h5);
+        }
+        
+        public static int Combine(int h1, int h2, int h3, int h4, int h5, int h6)
+        {
+            return Combine(Combine(h1, h2, h3, h4, h5), h6);
+        }
+        
+        public static int Combine(int h1, int h2, int h3, int h4, int h5, int h6, int h7)
+        {
+            return Combine(Combine(h1, h2, h3, h4, h5, h6), h7);
+        }
+        
+        public static int Combine(int h1, int h2, int h3, int h4, int h5, int h6, int h7, int h8)
+        {
+            return Combine(Combine(h1, h2, h3, h4, h5, h6, h7), h8);
+        }
+        
+        public static int CombineHashCodes<T1, T2>(T1 o1, T2 o2)
+        {
+            return Combine(GetHashCode(o1), GetHashCode(o2));
+        }
+        
+        public static int CombineHashCodes<T1, T2, T3>(T1 o1, T2 o2, T3 o3)
+        {
+            return Combine(GetHashCode(o1), GetHashCode(o2), GetHashCode(o3));
+        }
+        
+        public static int CombineHashCodes<T1, T2, T3, T4>(T1 o1, T2 o2, T3 o3, T4 o4)
+        {
+            return Combine(GetHashCode(o1), GetHashCode(o2), GetHashCode(o3), GetHashCode(o4));
+        }
+        
+        public static int CombineHashCodes<T1, T2, T3, T4, T5>(T1 o1, T2 o2, T3 o3, T4 o4, T5 o5)
+        {
+            return Combine(GetHashCode(o1), GetHashCode(o2),
+                GetHashCode(o3), GetHashCode(o4), GetHashCode(o5));
+        }
+        
+        public static int CombineHashCodes<T1, T2, T3, T4, T5, T6>(T1 o1, T2 o2, T3 o3, T4 o4, T5 o5, T6 o6)
+        {
+            return Combine(GetHashCode(o1), GetHashCode(o2),
+                GetHashCode(o3), GetHashCode(o4), GetHashCode(o5), GetHashCode(o6));
+        }
+        
+        public static int CombineHashCodes<T1, T2, T3, T4, T5, T6, T7>(T1 o1, T2 o2, T3 o3, T4 o4, T5 o5, T6 o6, T7 o7)
+        {
+            return Combine(GetHashCode(o1), GetHashCode(o2),
+                GetHashCode(o3), GetHashCode(o4), GetHashCode(o5),
+                GetHashCode(o6), GetHashCode(o7));
+        }
+        
+        public static int CombineHashCodes<T1, T2, T3, T4, T5, T6, T7, T8>(T1 o1, T2 o2, T3 o3, T4 o4, T5 o5, T6 o6, T7 o7, T8 o8)
+        {
+            return Combine(GetHashCode(o1), GetHashCode(o2),
+                GetHashCode(o3), GetHashCode(o4), GetHashCode(o5),
+                GetHashCode(o6), GetHashCode(o7), GetHashCode(o8));
+        }
+        
+        private static int GetHashCode<T>(T obj)
+        {
+            // Emulate the behavior of EqualityComparer<T>.Default,
+            // without introducing a virtual method call.
+            // For structs, it looks like the JIT will eliminate
+            // the null check here and simply call obj.GetHashCode().
+            return obj == null ? 0 : obj.GetHashCode();
```

Comment:
@omariom, this is a few months later, but... is enums the only case this penalizes for? I wonder if it would be a good trade off to box for enums, and avoid a virtual method call for all other cases.